### PR TITLE
add entrance for pagerank and wcc

### PR DESCRIPTION
**Pull request by zhoney**
Change-Id: I9d0a35ac8ea2cd7f1c9b6cf86a62a06e68d6498d

### CODE REVIEW

File: hugegraph-core/src/main/java/com/baidu/hugegraph/job/compute/WeakConnectedComponentCompute.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        return CATEGORY_COMM;
+    }
+
+    @Override
+    public void checkParameters(Map<String, Object> parameters) {
+        this.userDefinedParameters.put(MAX_STEPS, maxSteps(parameters));
+        this.userDefinedParameters.put(PRECISION, threshold(parameters));
+    }
+
+    private static double threshold (Map<String, Object> parameters) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/job/compute/WeakConnectedComponentCompute.java:49]: remove space in "threshold ("

File: hugegraph-core/src/main/java/com/baidu/hugegraph/job/compute/AbstractCompute.java [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+    public static Object parameter(Map<String, Object> parameters, String key) {
+        Object value = parameters.get(key);
+        E.checkArgument(value != null,
+                        "Expect '%s' in parameters: %s",
+                        key, parameters);
+        return value;
+    }
+
+    public static String parameterString(Map<String, Object> parameters,
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/job/compute/AbstractCompute.java:185]: add ParameterUtil class and move these methods to it

File: hugegraph-core/src/main/java/com/baidu/hugegraph/job/compute/AbstractCompute.java [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+        this.config = new YamlConfiguration();
+        this.config.load(configPath);
+
+        // Read common and compute specified parameters
+        this.commonConfig = this.readCommonConfig();
+    }
+
+    private Map<String, Object> readCommonConfig() {
+        return readSubConfig(COMMON);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/job/compute/AbstractCompute.java:107]: this.readSubConfig

File: hugegraph-core/src/main/java/com/baidu/hugegraph/job/ComputeJob.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        value = map.get("parameters");
+        E.checkArgument(value instanceof Map,
+                        "Invalid compute parameters '%s'", value);
+        @SuppressWarnings("unchecked")
+        Map<String, Object> parameters = (Map<String, Object>) value;
+
+        ComputePool pool = ComputePool.instance();
+        Compute compute = pool.find(name);
+        E.checkArgument(compute != null,
+                        "There is no compute named '%s'", name);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/job/ComputeJob.java:73]: There is no compute method named

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/job/ComputeAPI.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        }
+
+        HugeGraph g = graph(manager, graph);
+        Map<String, Object> input = ImmutableMap.of("compute", compute,
+                                                    "parameters", parameters);
+        JobBuilder<Object> builder = JobBuilder.of(g);
+        builder.name("compute:" + compute)
+               .input(JsonUtil.toJson(input))
+               .job(new ComputeJob());
+        return ImmutableMap.of("task_id", builder.schedule().id());
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/job/ComputeAPI.java:82]: define var for builder.schedule()

File: hugegraph-core/src/main/java/com/baidu/hugegraph/config/CoreOptions.java [hunk 6]
```diff
@@ -379,4 +379,12 @@ public static synchronized CoreOptions instance() {
                     disallowEmpty(),
                     "smart"
             );
+
+    public static final ConfigOption<String> COMPUTER_CONFIG =
+            new ConfigOption<>(
+                    "computer.config",
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/config/CoreOptions.java:385]: compute/computer unify into one

File: hugegraph-core/src/main/java/com/baidu/hugegraph/job/ComputeJob.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+import com.baidu.hugegraph.job.compute.ComputePool;
+import com.baidu.hugegraph.util.E;
+import com.baidu.hugegraph.util.JsonUtil;
+
+public class ComputeJob extends SysJob<Object> {
+
+    public static final String COMPUTE = "compute";
+
+    public static boolean check(String name, Map<String, Object> parameters) {
+        Compute algorithm = ComputePool.instance().find(name);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/job/ComputeJob.java:35]: rename

File: hugegraph-core/src/main/java/com/baidu/hugegraph/task/HugeServerInfo.java [hunk 8]
```diff
@@ -212,6 +212,10 @@ public static HugeServerInfo fromVertex(Vertex vertex) {
     }
 
     public <V> boolean suitableFor(HugeTask<V> task, long now) {
+        if (task.compute() && !this.role.computer() ||
+            !task.compute() && this.role.computer()) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/task/HugeServerInfo.java:216]: use task.compute() != this.role.computer()

File: hugegraph-core/src/main/java/com/baidu/hugegraph/job/compute/WeakConnectedComponentCompute.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        return CATEGORY_COMM;
+    }
+
+    @Override
+    public void checkParameters(Map<String, Object> parameters) {
+        this.userDefinedParameters.put(MAX_STEPS, maxSteps(parameters));
+        this.userDefinedParameters.put(PRECISION, threshold(parameters));
+    }
+
+    private static double threshold(Map<String, Object> parameters) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/job/compute/WeakConnectedComponentCompute.java:49]: rename to precision

File: hugegraph-core/src/main/java/com/baidu/hugegraph/job/compute/AbstractCompute.java [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+            results.put(node.getName(), extras);
+        }
+
+        return results;
+    }
+
+    @SuppressWarnings("unchecked")
+    private void setComputeSpecifiedParameters() {
+        ((Map<String, Object>) this.commonConfig.get(EXTRA_ARGS))
+                               .putAll(this.userDefinedParameters);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/job/compute/AbstractCompute.java:144]: align with "."

File: hugegraph-core/src/main/java/com/baidu/hugegraph/job/compute/PageRankCompute.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+    @Override
+    public String category() {
+        return CATEGORY_RANK;
+    }
+
+    @Override
+    public void checkParameters(Map<String, Object> parameters) {
+        this.userDefinedParameters.put(MAX_STEPS, maxSteps(parameters));
+        this.userDefinedParameters.put(ALPHA, alpha(parameters));
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/job/compute/PageRankCompute.java:46]: add precision

File: hugegraph-core/src/main/java/com/baidu/hugegraph/job/ComputeJob.java [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+        E.checkArgument(value instanceof Map,
+                        "Invalid compute parameters '%s'", value);
+        @SuppressWarnings("unchecked")
+        Map<String, Object> parameters = (Map<String, Object>) value;
+
+        ComputePool pool = ComputePool.instance();
+        Compute compute = pool.find(name);
+        E.checkArgument(compute != null,
+                        "There is no compute method named '%s'", name);
+        compute.checkParameters(parameters);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/job/ComputeJob.java:74]: move into call  and rename checkParameters to checkAndCollectParameters

File: hugegraph-core/src/main/java/com/baidu/hugegraph/job/compute/AbstractCompute.java [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+
+    public static final String MAX_STEPS = "max_steps";
+    public static final int DEFAULT_MAX_STEPS = 5;
+
+    protected static final String CATEGORY_RANK = "rank";
+    protected static final String CATEGORY_COMM = "community";
+
+    private YamlConfiguration config;
+    private Map<String, Object> commonConfig = new HashMap<>();
+    protected Map<String, Object> userDefinedParameters = new HashMap<>();
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/job/compute/AbstractCompute.java:58]: don't define as field

File: hugegraph-core/src/main/java/com/baidu/hugegraph/job/computer/AbstractComputer.java [hunk 14]
```diff
@@ -66,33 +66,35 @@ public void checkParameters(Map<String, Object> parameters) {
...
+            this.initializeConfig((ComputerJob) job);
         } catch (Exception e) {
             throw new HugeException(
                       "Failed to initialize computer config file", e);
         }
 
-        // Set current compute job's specified parameters
-        this.setComputeSpecifiedParameters();
+        // Set current computer job's specified parameters
+        this.commonConfig.putAll(this.checkAndCollectParameters(parameters));
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/job/computer/AbstractComputer.java:80]: don't update commonConfig

File: hugegraph-core/src/main/java/com/baidu/hugegraph/job/computer/AbstractComputer.java [hunk 15]
```diff
@@ -66,33 +66,35 @@ public void checkParameters(Map<String, Object> parameters) {
...
 
-        // Construct shell command for compute job
+        // Construct shell command for computer job
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/job/computer/AbstractComputer.java:82]: this.constructShellCommands

File: hugegraph-core/src/main/java/com/baidu/hugegraph/job/ComputerJob.java [hunk 16]
```diff
@@ -22,21 +22,21 @@
...
 import com.baidu.hugegraph.config.HugeConfig;
-import com.baidu.hugegraph.job.compute.Compute;
-import com.baidu.hugegraph.job.compute.ComputePool;
+import com.baidu.hugegraph.job.computer.Computer;
+import com.baidu.hugegraph.job.computer.ComputerPool;
 import com.baidu.hugegraph.util.E;
 import com.baidu.hugegraph.util.JsonUtil;
 
-public class ComputeJob extends SysJob<Object> {
+public class ComputerJob extends SysJob<Object> {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/job/ComputerJob.java:30]: extends UserJob

Comment (zhoney) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/job/ComputerJob.java:30]: need get computer conf from SysJob's params

File: hugegraph-core/src/main/java/com/baidu/hugegraph/job/ComputerJob.java [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+    public static boolean check(String name, Map<String, Object> parameters) {
+        Computer computer = ComputerPool.instance().find(name);
+        if (computer == null) {
+            return false;
+        }
+        computer.checkParameters(parameters);
+        return true;
+    }
+
+    public HugeConfig config() {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/job/ComputerJob.java:43]: do not expose HugeConfig  to the outside

File: hugegraph-core/src/main/java/com/baidu/hugegraph/job/computer/AbstractComputer.java [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+    private static final String COMMAND_PREFIX =
+            "cd $COMPUTER_HOME;" +
+            "hadoop jar hugegraph-computer.jar com.baidu.hugegraph.Computer " +
+            "-D libjars=./hugegraph-computer-core.jar";
+
+    private static final String COMMON = "common";
+    private static final String ARG_SYMBOL = "C";
+    private static final String MINUS = "-";
+    private static final String EQUAL = "=";
+    private static final String EMPTY = " ";
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/job/computer/AbstractComputer.java:47]: SPACE

File: hugegraph-core/src/main/java/com/baidu/hugegraph/job/computer/AbstractComputer.java [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+
+        // Construct shell command for computer job
+        String command = this.constructShellCommands(configs);
+
+        // Execute current computer
+        int exitCode;
+        try {
+            Process process = Runtime.getRuntime().exec(command);
+            exitCode = process.waitFor();
+        } catch (Exception e) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/job/computer/AbstractComputer.java:92]: cache throwable

File: hugegraph-core/src/main/java/com/baidu/hugegraph/job/computer/AbstractComputer.java [hunk 20]
```diff
@@ -0,0 +1,170 @@
...
+                        MAX_STEPS, maxSteps);
+        return maxSteps;
+    }
+
+    protected static double precision(Map<String, Object> parameters) {
+        if (!parameters.containsKey(PRECISION)) {
+            return DEFAULT_PRECISION;
+        }
+        double precision = ParameterUtil.parameterDouble(parameters, PRECISION);
+        E.checkArgument(precision > 0 && precision < 1,
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/job/computer/AbstractComputer.java:165]: change to 0.0 1.0

File: hugegraph-dist/src/assembly/static/conf/computer.yaml [hunk 21]
```diff
@@ -0,0 +1,16 @@
+# common parameters
+common : {
+  worker_heap: 10000,
+  vertex_input_dir: /data/data_wdc_600/arc,
+  workers: 51,
+  zookeeper_list: "10.157.11.170:2181",
```
Comment (javeme) [inline hugegraph-dist/src/assembly/static/conf/computer.yaml:6]: 127.0

**Pull request merged**